src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala

```diff
@@ -218,50 +257,7 @@ private[spark] case class CosmosDBConnection(config: Config) extends CosmosDBLog
           }
 
           val error:DocumentClientException = outerError.getCause().asInstanceOf[DocumentClientException]
-          if (counter > 1000 ||
-            error.getStatusCode() != 410) {
-
-            throw outerError
-          }
-
-          if (counter > 0) {
-            val delay = CosmosDBConnection.getRandomRetryInterval(counter)
-            logInfo(s"Retrying query after Invalid Partition Key error with delay of $delay ms")
-            Thread.sleep(delay)
-          } else {
-            logInfo(s"Retrying query after Invalid Partition Key error without delay")
-          }
-
-          counter = counter + 1
-
-          CosmosDBConnectionCache.purgeCache(clientConfig)
-          val pkRanges = this.getAllPartitions
-          if (pkRanges
-            .find((pkRange) => pkRange.getId.equalsIgnoreCase(originalFeedOptions.getPartitionKeyRangeIdInternal))
-            .isDefined) {
-
-            childPKRanges = Array.empty[String]
-          } else {
-            // NOTE - the partition that was passed into the original FeedOptions does not exist anymore
-            // Which means a partition split completed since Spark did the partitioning for the query
-            // below code tries to identify the child partitions and the next retry will create a chained iterator
-            // for the queries on these child partitions instead of trying to query the original partition again
-            // WARNING!!! - this will only work for partition splits - not merges. That is fine because V2 Java SDK
-            // is neither partition split proof and definitely not merge proof - it has always clearly been communicated
-            // that the Spark 2.4 connector will not allow for partition merge - only the new Spark 3 connector will be
-            // able to handle merges.
-            val newChildRanges : mutable.Buffer[String] = new mutable.ArrayBuffer[String]()
-            pkRanges.foreach((pkRange) => {
-              if (pkRange.getParents != null &&
-                pkRange.getParents.contains(originalFeedOptions.getPartitionKeyRangeIdInternal)) {
-
-                newChildRanges += pkRange.getId()
-                logInfo(s"Found child partition '${pkRange.getId}' for original " +
-                  s"partition '${originalFeedOptions.getPartitionKeyRangeIdInternal}'.")
-              }
-            })
-            childPKRanges = newChildRanges.toArray
-          }
+          tryHandlePartitionSplitError(outerError, error, originalFeedOptions, counter)
```

Comment:
nit: Should we be updating the value of childRanges with the output of the call?

Reply:
Yes - thanks!

Reply:
Fixed in next iteration